src/wallet/wallet.cpp

```diff
@@ -1293,21 +1293,39 @@ bool CWallet::AddToWallet(const CWalletTx& wtxIn, bool fFromLoadWallet, CWalletD
 
 bool CWallet::UpdatedNoteData(const CWalletTx& wtxIn, CWalletTx& wtx)
 {
-    if (wtxIn.mapSproutNoteData.empty() || wtxIn.mapSproutNoteData == wtx.mapSproutNoteData) {
-        return false;
+    bool unchangedSproutFlag = (wtxIn.mapSproutNoteData.empty() || wtxIn.mapSproutNoteData == wtx.mapSproutNoteData);
+    if (!unchangedSproutFlag) {
+        auto tmp = wtxIn.mapSproutNoteData;
+        // Ensure we keep any cached witnesses we may already have
+        for (const std::pair <JSOutPoint, SproutNoteData> nd : wtx.mapSproutNoteData) {
+            if (tmp.count(nd.first) && nd.second.witnesses.size() > 0) {
+                tmp.at(nd.first).witnesses.assign(
+                        nd.second.witnesses.cbegin(), nd.second.witnesses.cend());
+            }
+            tmp.at(nd.first).witnessHeight = nd.second.witnessHeight;
+        }
+        // Now copy over the updated note data
+        wtx.mapSproutNoteData = tmp;
     }
-    auto tmp = wtxIn.mapSproutNoteData;
-    // Ensure we keep any cached witnesses we may already have
-    for (const std::pair<JSOutPoint, SproutNoteData> nd : wtx.mapSproutNoteData) {
-        if (tmp.count(nd.first) && nd.second.witnesses.size() > 0) {
-            tmp.at(nd.first).witnesses.assign(
-                nd.second.witnesses.cbegin(), nd.second.witnesses.cend());
+
+    bool unchangedSaplingFlag = (wtxIn.mapSaplingNoteData.empty() || wtxIn.mapSaplingNoteData == wtx.mapSaplingNoteData);
+    if (!unchangedSaplingFlag) {
+        auto tmp = wtxIn.mapSaplingNoteData;
+        // Ensure we keep any cached witnesses we may already have
+
+        for (const std::pair <SaplingOutPoint, SaplingNoteData> nd : wtx.mapSaplingNoteData) {
+            if (tmp.count(nd.first) && nd.second.witnesses.size() > 0) {
+                tmp.at(nd.first).witnesses.assign(
+                        nd.second.witnesses.cbegin(), nd.second.witnesses.cend());
+            }
+            tmp.at(nd.first).witnessHeight = nd.second.witnessHeight;
         }
-        tmp.at(nd.first).witnessHeight = nd.second.witnessHeight;
+
+        // Now copy over the updated note data
+        wtx.mapSaplingNoteData = tmp;
     }
-    // Now copy over the updated note data
-    wtx.mapSproutNoteData = tmp;
-    return true;
+
+    return !unchangedSproutFlag || !unchangedSaplingFlag;
```

Comment:
This could be written as a template method:
```
template<typename OutPoint, typename NoteData>
bool UpdateNoteData(const std::map<OutPoint, NoteData>& wtxInMap, std::map<OutPoint, NoteData>& wtxMap)
{
    if (wtxInMap.empty() || wtxInMap == wtxMap) {
        return false;
    }
    auto tmp = wtxInMap;
    // Ensure we keep any cached witnesses we may already have
    for (const std::pair <OutPoint, NoteData> nd : wtxMap) {
        if (tmp.count(nd.first) && nd.second.witnesses.size() > 0) {
            tmp.at(nd.first).witnesses.assign(
                    nd.second.witnesses.cbegin(), nd.second.witnesses.cend());
        }
        tmp.at(nd.first).witnessHeight = nd.second.witnessHeight;
    }
    // Now copy over the updated note data
    wtxMap = tmp;
    return true;
}

bool CWallet::UpdatedNoteData(const CWalletTx& wtxIn, CWalletTx& wtx)
{
    bool changedSprout = ::UpdateNoteData(wtxIn.mapSproutNoteData, wtx.mapSproutNoteData);
    bool changedSapling = ::UpdateNoteData(wtxIn.mapSaplingNoteData, wtx.mapSaplingNoteData);
    return changedSprout || changedSapling;
}
```
I prefer this because it prevents future divergence and puts one step closer to an abstraction that encapsulates logic such as this.

Reply:
Can be done in a future PR.